src/main/webapp/app/guided-tour/guided-tour.service.ts

```diff
@@ -0,0 +1,371 @@
+        if (this.currentTour.steps[this.currentTourStepIndex + 1]) {
+            this.currentTourStepIndex++;
+            if (currentStep.action) {
+                currentStep.action();
+            }
+            // Usually an action is opening something so we need to give it time to render.
+            setTimeout(() => {
+                if (this.checkSelectorValidity()) {
+                    this.guidedTourCurrentStepSubject.next(this.getPreparedTourStep(this.currentTourStepIndex));
+                } else {
+                    this.nextStep();
+                }
+            });
+        } else {
+            if (this.currentTour.completeCallback) {
+                this.currentTour.completeCallback();
+            }
+            this.updateGuidedTourSettings(this.currentTour.settingsId, false).subscribe(guidedTourSettings => {
+                if (guidedTourSettings.body) {
+                    this.guidedTourSettings = guidedTourSettings.body;
+                }
+            });
+            this.resetTour();
+        }
+    }
+
+    /**
+     * Navigate to previous tour step
+     */
+    public backStep(): void {
+        if (this.currentTour) {
+            const currentStep = this.currentTour.steps[this.currentTourStepIndex];
+            if (currentStep.closeAction) {
+                currentStep.closeAction();
+            }
+            if (this.currentTour.steps[this.currentTourStepIndex - 1]) {
+                this.currentTourStepIndex--;
+                if (currentStep.action) {
+                    currentStep.action();
+                }
+                setTimeout(() => {
+                    if (this.checkSelectorValidity()) {
+                        this.guidedTourCurrentStepSubject.next(this.getPreparedTourStep(this.currentTourStepIndex));
+                    } else {
+                        this.backStep();
+                    }
+                });
+            } else {
+                this.resetTour();
+            }
+        }
+    }
+
+    /**
+     * Skip tour and exit the tour context
+     */
+    public skipTour(): void {
+        if (this.currentTour) {
+            if (this.currentTour.skipCallback) {
+                this.currentTour.skipCallback(this.currentTourStepIndex);
+            }
+            this.resetTour();
+        }
+    }
+
+    /**
+     * Close tour by resetting `currentTour`, `currentTourStepIndex` and `guidedTourCurrentStepSubject`
+     * and remove overlay
+     */
+    public resetTour(): void {
+        document.body.classList.remove('tour-open');
+        this.currentTour = null;
+        this.currentTourStepIndex = 0;
+        this.guidedTourCurrentStepSubject.next(null);
+    }
+
+    /**
+     * Start guided tour for given guided tour
+     * @param {tour} guided tour
+     */
+    private startTour(tour: GuidedTour): void {
+        this.currentTourSteps = tour.steps;
+
+        // adjust tour steps according to permissions
+        tour.steps.forEach((step, index) => {
```

Comment:
This could also be a filter operation:

`this.currentTourSteps = tour.steps.filter(step => !step.permission || this.accountService.hasAnyAuthorityDirect(step.permission));`

To me this is a lot clearer to what it means and it does not mutate the existing array.

Reply:
@thilo-behnke Thanks for the suggestion! I adapted the code as such.